src/Instrumentation/Slim/src/SlimInstrumentation.php

```diff
@@ -0,0 +1,148 @@
+                $parent = Context::getCurrent();
+                $builder = $instrumentation->tracer()->spanBuilder(sprintf('HTTP %s', $request?->getMethod() ?? 'unknown'))
+                    ->setSpanKind(SpanKind::KIND_SERVER)
+                    ->setAttribute('code.function', $function)
+                    ->setAttribute('code.namespace', $class)
+                    ->setAttribute('code.filepath', $filename)
+                    ->setAttribute('code.lineno', $lineno);
+                if ($request) {
+                    //extract trace propagation headers from request
+                    $parent = Globals::propagator()->extract($request->getHeaders());
+
+                    $builder = $builder
+                        ->setAttribute(TraceAttributes::HTTP_URL, $request->getUri()->__toString())
+                        ->setAttribute(TraceAttributes::HTTP_METHOD, $request->getMethod())
+                        ->setAttribute(TraceAttributes::HTTP_REQUEST_CONTENT_LENGTH, $request->getHeaderLine('Content-Length'))
+                        ->setAttribute(TraceAttributes::HTTP_SCHEME, $request->getUri()->getScheme());
+                }
+                $span = $builder->setParent($parent)->startSpan();
+
+                Context::storage()->attach($span->storeInContext($parent));
+                if ($request) {
+                    return [$request->withAttribute(SpanInterface::class, $span)];
+                }
+
+                return [];
+            },
+            post: static function (RequestHandlerInterface $handler, array $params, ?ResponseInterface $response, ?Throwable $exception) {
+                $scope = Context::storage()->scope();
+                if (!$scope) {
+                    return;
+                }
+                $scope->detach();
+                $span = Span::fromContext($scope->context());
+                if ($exception) {
+                    $span->recordException($exception, [TraceAttributes::EXCEPTION_ESCAPED => true]);
+                    $span->setStatus(StatusCode::STATUS_ERROR, $exception->getMessage());
+                }
+                if ($response?->getStatusCode() >= 500) {
+                    $span->setStatus(StatusCode::STATUS_ERROR);
+                }
+
+                $span->end();
+            }
+        );
+
+        /**
+         * Update root span's name after Slim routing, using either route name or method+pattern.
+         * This relies upon the existence of a request attribute with key SpanInterface::class
+         * and type SpanInterface which represents the root span, having been previously set (eg
+         * by Psr15 auto-instrumentation)
+         *
+         * @psalm-suppress ArgumentTypeCoercion
+         */
+        hook(
+            RoutingMiddleware::class,
+            'performRouting',
+            pre: null,
+            post: static function (RoutingMiddleware $middleware, array $params, ServerRequestInterface $request, ?Throwable $exception) {
```

Comment:
This hook is used only to update span name created by other hook. I think that hooks should be self contained and orthogonal. I would remove this hook and use `request->getUri()` in `handle` hook(s)

Reply:
The reason I've done it as <set> then later <update> is:
- the spec says that span names by default should not be the uri: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name (too-high cardinality)
- slim doesn't have a better route name until after the routing middleware has run (which it might not do, if an error occurs earlier)

Reply:
Maybe, we can extract part of uri, it would simplify code.